### ISSUE TITLE
Could pressing "Enter" with no text typed be ignored rather than counting as an attempt?

### ISSUE THREAD

**Issue by whereswaldon** [ifreund/waylock#62]
This may be the wrong layer to address this at, but I keep finding that I'm losing a password attempt when I wake my computer up if I press the "Enter" key to do so (old habit). Since I haven't typed any text at all prior to that, it seems kinda silly for `waylock` to attempt the current empty string as a password input. I have a complex random password, and this mistake combined with typos has gotten me locked out of my account for 10 minutes a few times.

I guess this is the interaction of the default Arch Linux PAM config and waylock. I could give myself more than 3 attempts, but 3 does seem like a good number.

Are you open to allowing waylock to ignore this scenario? Either way, thanks a bunch for your work on this project and the Wayland ecosystem!

**Comment by ifreund** [comment]
Hmm, the only thing I'm worried about is whether or not unixy systems consider the empty string to be a valid password. Say your password was for some reason the empty string, if so then waylock ignoring the enter key with no password input would lock you out of the system.

I'm inclined to believe that even if such a configuration is valid, such a user has no use for a screenlocker in the first place though and it's probably reasonable for waylock to ignore attempts to submit an empty password.

In any case, the patch to implement this is trivial:
```diff
diff --git a/src/Lock.zig b/src/Lock.zig
index a54d038..6b9f8a0 100644
--- a/src/Lock.zig
+++ b/src/Lock.zig
@@ -381,6 +381,11 @@ fn session_lock_listener(_: *ext.SessionLockV1, event: ext.SessionLockV1.Event,
 pub fn submit_password(lock: *Lock) void {
     assert(lock.state == .locked);
 
+    if (lock.password.buffer.len == 0) {
+        log.info("ignoring submission of empty password", .{});
+        return;
+    }
+
     lock.send_password_to_auth() catch |err| {
         fatal("failed to send password to child authentication process: {s}", .{@errorName(err)});
     };
```

**Comment by whereswaldon** [comment]
It seems like you can go out of your way to set the password of a Linux system to the empty string by storing the pre-computed hash of the empty string: https://askubuntu.com/a/1080575

The rest of the search hits for this warn about things like systems that default-deny users with no password set. I don't really know how to get a definitive answer on this, but I would tend to agree that a user with no password also needs no screenlocker.

Thank you for the patch! Do you plan to commit it, or should I just run a patched version indefinitely?

**Comment by ifreund** [comment]
> Thank you for the patch! Do you plan to commit it, or should I just run a patched version indefinitely?

I think I'll commit it, I just wanted to take a peek at some other software first to see if anyone else (say xscreensaver for example) is already doing this.

**Comment by whereswaldon** [comment]
Okay, thanks for looking into it!

**Comment by phrom** [comment]
swaylock does this, but it's opt-in through a --ignore-empty-password config option.

**Comment by Nyabinary** [comment]
What the status of this?

**Comment by ifreund** [comment]
From this commit message it sounds like there are valid reasons to allow submission of an empty password: https://github.com/i3/i3lock/commit/5627a111746e96119e6789d8f8f47168b3a8c41d

> Some users may want to validate an empty password: PAM may rely on
> other sources to unlock the screen, like the presence of a token or
> the proximity of some Bluetooth device. Hence, we don't forbid this
> possibility and provide an command-line option for users not willing
> to validate empty passwords.

I think we therefore should make this configurable, the only remaining question is whether it should be opt-in or opt-out. 
Since ignoring empty passwords apparently has the possibility of locking out users which currently use waylock with some kind of alternative pam authentication and since it would be a behavior change I think the status quo is the correct default.

I'll add a `-ignore-empty-password` flag shortly.

**Comment by ifreund** [comment]
Implemented in b9bd798d5182ccc26294437e4cfdf3e728562f56

**Comment by whereswaldon** [comment]
Thank you @ifreund! The attention to detail in your implementation choice is appreciated.
